molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+                        'instances'][index]:
+                    self.molecule.config.config['libvirt']['instances'][index][
+                        'image'].update(image)
+                else:
+                    self.molecule.config.config['libvirt']['instances'][index][
+                        'image'] = image
+        return image
+
+    def up(self, no_provision=True):
+        """
+        Up or define/up libvirt instances.
+
+        :param: no_provision is not meaningful for libvirt instances
+        """
+        for net in self.molecule.config.config['libvirt']['networks']:
+            self._up_network(net)
+        pool = self._up_pool()
+        vols = pool.listAllVolumes()
+        domains = self._libvirt.listAllDomains()
+        wait = True
+        for instance in self.instances:
+            image = self._populate_image(instance)
+            imagefile = image['source'].split('/')[-1]
+            # Ensure that the image is available
+            if not os.path.exists(
+                    os.path.join(self._pool_path, image['name'] + '.img')):
+                if not os.path.exists(
+                        os.path.join(self._sources_path, imagefile)):
+                    self._fetch(image['source'], imagefile)
+                if imagefile.endswith('.box'):
+                    self._unpack_box(image, imagefile)
+
+            # Is there an existing libvirt volume for this instance?
+            try:
+                vol_found = pool.storageVolLookupByName(instance['name'])
+                if vol_found and not vol_found.isActive():
+                    vol_found.create()
+                return vol_found  # existing volume is now running
+            except libvirt.libvirtError as e:
+                self._create_volume(pool, instance)
+            # TODO: Are all this instance's interface's networks available?
+            # Is there an existing libvirt domain defined for this instance?
+            dom_found = False
+            for dom in domains:
+                if not dom_found:
+                    if dom.name() == instance['name']:
+                        dom_found = True
+                        if dom.info()[0] == 1:
+                            utilities.print_info(
+                                "\t{}: already running".format(instance[
+                                    'name']))
+                            break
+                        else:
+                            utilities.print_info("\t{}: booting".format(
+                                instance['name']))
+                            dom.create()
+                            if wait:
+                                time.sleep(self._boot_wait)
+                                wait = False
+            if not dom_found:
+                utilities.print_info("\t{}: defining".format(instance['name']))
+                dom = self._libvirt.defineXML(self._build_domain_xml(instance))
+                if GUESTFS:
```

Comment:
Is it reasonable to keep the "no guestfs available" path?  It complicates things.
If you really want to keep it, I suggest adding an escape-style path instead of a test, to limit the nested block:
```
if not GUESTFS:
    return
# Continue normal flow.
```